smac/intensification/intensification.py

```diff
@@ -137,94 +190,362 @@ def __init__(self, tae_runner: ExecuteTARun,
+        if self.stage in [IntensifierStage.RUN_FIRST_CONFIG, IntensifierStage.RUN_INCUMBENT]:
+
+            # Line 3
+            # A modified version, that not only checks for maxR
+            # but also makes sure that there are runnable instances,
+            # that is, instances has not been exhausted
+            inc_runs = run_history.get_runs_for_config(
+                incumbent,
+                only_max_observed_budget=True
+            )
+
+            # Line 4
+            available_insts = self._get_inc_available_inst(incumbent, run_history)
+            if available_insts and len(inc_runs) < self.maxR:
+                # Lines 5-6-7
+                instance, seed, cutoff = self._get_next_inc_config(available_insts)
+
+                instance_specific = "0"
+                if instance is not None:
+                    instance_specific = self.instance_specifics.get(instance, "0")
+
+                return RunInfoIntent.RUN, RunInfo(
+                    config=incumbent,
+                    instance=instance,
+                    instance_specific=instance_specific,
+                    seed=seed,
+                    cutoff=cutoff,
+                    capped=False,
+                    budget=0.0,
+                )
+            else:
+                # This point marks the transitions from lines 3-7
+                # to 8-18.
+
+                self.logger.debug(
+                    "No further instance-seed pairs for incumbent available."
+                )
+
+                self.stage = IntensifierStage.RUN_CHALLENGER
+
+                self._process_inc_run(
+                    incumbent=incumbent,
+                    run_history=run_history,
+                )
+
+        # Understand who is the active challenger.
+        if self.stage == IntensifierStage.RUN_BASIS:
+            # if in RUN_BASIS stage,
+            # return the basis configuration (i.e., `always_race_against`)
+            self.logger.debug("Race against basis configuration after incumbent change.")
+            challenger = self.always_race_against
+        elif self.current_challenger and self.continue_challenger:
+            # if the current challenger could not be rejected,
+            # it is run again on more instances
+            challenger = self.current_challenger
+        else:
+            # Get a new challenger if all instance/pairs have
+            # been completed. Else return the currently running
+            # challenger
+            challenger, self.new_challenger = self.get_next_challenger(
+                challengers=challengers,
+                chooser=chooser,
+            )
+
+        # No new challengers are available for this iteration,
+        # Move to the next iteration. This can only happen
+        # when all configurations for this iteration are exhausted
+        # and have been run in all proposed instance/pairs.
+        if challenger is None:
+            return RunInfoIntent.SKIP, RunInfo(
```

Comment:
Don't you also have to call `_next_iteration`?

Reply:
As in the previous code, this is something that get_next_challenger does.

I think it makes sense to keep it there because it highlights when the iteration is done?

If you think otherwise, I can move it here.

Reply:
Okay, no I get it - could you please add a not to the docstring of `get_next_challenger` that it can trigger a new iteration?

Reply:
thanks, I will update the comment to be:
This method triggers a call to _next_iteration if there are no more configurations
           to run, for the current intensification loop. This marks the transition to Line 2,
           where a new configuration to intensify will be drawn from epm/initial challengers.